lib/ClangImporter/ImportType.cpp

```diff
@@ -2230,6 +2230,157 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+    const clang::ParmVarDecl *param, OptionalTypeKind optionalityOfParam,
+    ImportTypeKind importKind, bool allowNSUIntegerAsInt,
+    bool isNSDictionarySubscriptGetter, bool paramIsError,
+    bool paramIsCompletionHandler,
+    Optional<unsigned> completionHandlerErrorParamIndex,
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn,
+    bool &isInOut, bool &isParamTypeImplicitlyUnwrapped) {
+  auto paramTy = param->getType();
+
+  // Import the parameter type into Swift.
+  auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
+  Type swiftParamTy;
+
+  // Sometimes we import unavailable typedefs as enums. If that's the case,
+  // use the enum, not the typedef here.
+  if (auto typedefType = dyn_cast<clang::TypedefType>(paramTy.getTypePtr())) {
+    if (isUnavailableInSwift(typedefType->getDecl())) {
+      if (auto clangEnum =
+              findAnonymousEnumForTypedef(SwiftContext, typedefType)) {
+        // If this fails, it means that we need a stronger predicate for
+        // determining the relationship between an enum and typedef.
+        assert(clangEnum.getValue()
+                   ->getIntegerType()
+                   ->getCanonicalTypeInternal() ==
+               typedefType->getCanonicalTypeInternal());
+        if (auto swiftEnum = importDecl(*clangEnum, CurrentVersion)) {
+          swiftParamTy = cast<NominalTypeDecl>(swiftEnum)->getDeclaredType();
+        }
+      }
+    }
+  } else if (isa<clang::PointerType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto pointeeType = paramTy->getPointeeType();
+    auto templateParamType = cast<clang::TemplateTypeParmType>(pointeeType);
+    PointerTypeKind pointerKind = pointeeType.getQualifiers().hasConst()
+                                      ? PTK_UnsafePointer
+                                      : PTK_UnsafeMutablePointer;
+    auto genericType = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+    swiftParamTy = genericType->wrapInPointer(pointerKind);
+    if (!swiftParamTy)
+      return None;
+  } else if (isa<clang::ReferenceType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto templateParamType =
+        cast<clang::TemplateTypeParmType>(paramTy->getPointeeType());
+    swiftParamTy = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+    if (!paramTy->getPointeeType().isConstQualified())
+      isInOut = true;
+  } else if (auto *templateParamType =
+                 dyn_cast<clang::TemplateTypeParmType>(paramTy)) {
+    swiftParamTy = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+  } else if (auto refType = dyn_cast<clang::ReferenceType>(paramTy)) {
```

Comment:
@plotfi this is the branch that was missing in the Obj-C++ case. From line 2276 to 2304 it did not exist for methods. The branches before this are not taken because they depend on templates, but this branch checks if the `paramTy` is a reference type (which is the case for C++ structs/class), and allows using it without the `UnsafePointer`. Since `paramTy` is changed to the pointee, the later code that calls `importType` does not add the `UnsafePointer`.

PD: I don't know why I cannot comment this in the commit itself, but this branch does not appear in the commit in question because it just modifies the code around this, not this code directly.